Sources/ConsoleKit/Utilities/LoggerFragment.swift

```diff
@@ -0,0 +1,271 @@
+import Logging
+
+/// Information about a specific log message, including information from the logger the message was logged to.
+public struct LogRecord {
+	/// The log level of the message
+	public var level: Logger.Level
+	/// The logged message
+	public var message: Logger.Message
+	/// The metadata explicitly associated with the logged message
+	public var metadata: Logger.Metadata?
+	/// The source of the log message, usually the module name
+	public var source: String
+	/// The file the message was logged from
+	public var file: String
+	/// The function the message was logged from
+	public var function: String
+	/// The line number in the file the message was logged from
+	public var line: UInt
+	
+	
+	/// The label of the logger the message was logged to
+	public var label: String
+	/// The log level of the logger the message was logged to
+	public var loggerLevel: Logger.Level
+	/// The metadata associated with the logger the message was logged to
+	public var loggerMetadata: Logger.Metadata
+	/// The metadata provider associated with the logger the message was logged to
+	public var metadataProvider: Logger.MetadataProvider?
+	
+	/// The resolved metadata, combining all of the sources. This is computed lazily.
+	var _allMetadata: [String: Logger.MetadataValue]? = nil
+	
+	/// Combine all of the metadata into a single set. The result is cached after it is computed once.
+	public mutating func allMetadata() -> [String: Logger.MetadataValue] {
+		if let all = self._allMetadata {
+			return all
+		} else {
+			let meta = (self.metadata ?? [:])
+				.merging(self.loggerMetadata, uniquingKeysWith: { (a, _) in a })
+				.merging(self.metadataProvider?.get() ?? [:], uniquingKeysWith: { (a, _) in a })
+			self._allMetadata = meta
+			return meta
+		}
+	}
```

Comment:
Normally I'd very much applaud the caching here, but in practice does this ever get evaluated more than once per record? (And if it _does_, then the output of the `metadataProvider` could have changed, making the cache invalid - as unlikely as that probably is.)

Reply:
I was thinking it might be common-ish for end users to want to use metadata to filter out messages first thing in the pipeline. In that case the metadata would have to be constructed twice for every message that was actually logged, once to determine whether to filter, and once to actually output it.

Maybe that's not a use case worth optimizing for though.

Reply:
I'd leave it out for now, but make sure you do so in a way that allows putting it back later if needed _without_ breaking public API.